packages/adapter-edgedb/src/index.ts

```diff
@@ -0,0 +1,301 @@
+import type { Adapter, AdapterSession, AdapterUser } from "next-auth/adapters";
+import type { Client } from "edgedb";
+
+export function EdgeDBAdapter(client: Client): Adapter {
+  return {
+    async createUser({ email, emailVerified, name, image }) {
+      return await client.queryRequiredSingle(
+        `
+        with
+          image := <optional str>$image,
+          name := <optional str>$name,
+          emailVerified := <optional str>$emailVerified
+
+        select (
+          insert User {
+            email:= <str>$email,
+            emailVerified:= <datetime>emailVerified,
+            name:= name ?? .name,
+            image:= image ?? .image,
+          }
+        ) {
+            id,
+            email,
+            emailVerified,
+            name,
+            image
+          }
+        `,
+        { email, emailVerified: emailVerified && new Date(emailVerified).toISOString(), name, image }
+      );
+    },
+    async getUser(id) {
+      return await client.querySingle(
+        `
+        select User {
+          id,
+          email,
+          emailVerified,
+          name,
+          image
+        } filter .id = <uuid>$id;
+        `,
+        { id }
+      );
+    },
+    async getUserByEmail(email) {
+      return await client.querySingle(
+        `
+        select User {
+          id,
+          email,
+          emailVerified,
+          name,
+          image
+        } filter .email = <str>$email;
+        `,
+        { email }
+      );
+    },
+    async getUserByAccount({ providerAccountId, provider }) {
+      return await client.querySingle(
+        `
+        select User {
+          id,
+          email,
+          image,
+          name,
+          emailVerified,
+        } filter 
+          .accounts.providerAccountId = <str>$providerAccountId
+          and
+          .accounts.provider = <str>$provider
+          limit 1;
```

Comment:
Wrapping the query in `assert_single(...)` instead of using `limit 1` is the recommended best practice here, because `limit 1` will just silently discard results, instead of failing if the query does somehow return more than one `User`. Though since there's an exclusive constraint on `(.provider, .providerAccountId)`, querying the `Account` first then selecting the `user` link on it should let EdgeDB infer it's single without the need for `assert_single`. eg.
```
with account := (
  select Account
  filter .providerAccountId = <str>$providerAccountId
     and .provider = <str>$provider
)
select account.user {
  id,
  email,
  image,
  name,
  emailVerified
}
```